extension/agenthealth/handler/stats/provider/statuscode.go

```diff
@@ -0,0 +1,173 @@
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/amazon-contributing/opentelemetry-collector-contrib/extension/awsmiddleware"
+
+	"github.com/aws/amazon-cloudwatch-agent/extension/agenthealth/handler/stats/agent"
+)
+
+const (
+	statusResetInterval = 5 * time.Minute
+	statusHandlerID     = "cloudwatchagent.StatusCodeHandler"
+)
+
+var (
+	statusCodeProviderSingleton *StatusCodeProvider
+	StatusCodeProviderOnce      sync.Once
+)
+
+// StatusCodeProvider handles processing of status codes and maintains stats.
+type StatusCodeProvider struct {
+	statsByOperation map[string]*[5]int
+	resetTimer       *time.Timer
+	statusCodeChan   chan statusCodeEntry
+	stopChan         chan struct{}
+	shouldResetStats bool
+	mu               sync.Mutex
+	wg               sync.WaitGroup
+}
+
+// statusCodeEntry represents a status code and its associated operation.
+type statusCodeEntry struct {
+	operation  string
+	statusCode int
+}
+
+// GetStatusCodeStatusCodeProvider initializes and retrieves the singleton StatusCodeProvider.
+func GetStatusCodeStatsProvider() *StatusCodeProvider {
+	StatusCodeProviderOnce.Do(func() {
+		provider := &StatusCodeProvider{
+			statsByOperation: make(map[string]*[5]int),
+			statusCodeChan:   make(chan statusCodeEntry),
+			stopChan:         make(chan struct{}),
+		}
+
+		provider.startResetTimer()
+		provider.startProcessing()
+		statusCodeProviderSingleton = provider
+	})
+	return statusCodeProviderSingleton
+}
+
+// startProcessing begins processing status codes from the channel.
+func (sp *StatusCodeProvider) startProcessing() {
+	sp.wg.Add(1)
+	go func() {
+		defer sp.wg.Done()
+		for {
+			select {
+			case entry := <-sp.statusCodeChan:
+				sp.processStatusCode(entry)
+			case <-sp.stopChan:
+				return
+			}
+		}
+	}()
+}
+
+// EnqueueStatusCode adds a status code entry to the channel.
+func (sp *StatusCodeProvider) EnqueueStatusCode(operation string, statusCode int) {
+	sp.statusCodeChan <- statusCodeEntry{operation: operation, statusCode: statusCode}
+}
+
+// processStatusCode updates the stats map for the given status code entry.
+func (sp *StatusCodeProvider) processStatusCode(entry statusCodeEntry) {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+
+	stats, exists := sp.statsByOperation[entry.operation]
+	if !exists {
+		stats = &[5]int{}
+		sp.statsByOperation[entry.operation] = stats
+	}
+	sp.updateStatusCodeCount(stats, entry.statusCode)
+}
+
+// updateStatusCodeCount updates the count for the specific status code.
+func (sp *StatusCodeProvider) updateStatusCodeCount(stats *[5]int, statusCode int) {
+	switch statusCode {
+	case 200:
+		stats[0]++
+	case 400:
+		stats[1]++
+	case 408:
+		stats[2]++
+	case 413:
+		stats[3]++
+	case 429:
+		stats[4]++
+	default:
+		return
+	}
+}
+
+// startResetTimer initializes a reset timer to clear stats periodically.
+func (sp *StatusCodeProvider) startResetTimer() {
+	sp.resetTimer = time.AfterFunc(statusResetInterval, func() {
+		sp.shouldResetStats = true
```

Comment:
May have a data race where the bool is getting updated at the same time as it's being read. Please add a test for this reset.